File: xaiop-sdk/nodejs/src/core/history.ts
// @ts-nocheck
/**
 * Optional parse-chain history (flight recorder) for `.` phase boundaries.
 *
 * Two independent modes (both default **off**):
 * - **Snapshot** — read-only cursor: export time-root, range view, compare, URL lifecycle.
 * - **Realtime** — live jump: keep positioning node, permanently discard everything after.
 *
 * Both may be enabled together: inspect via snapshot, then cut the live sequence with realtime.
 */

import { cloneJson } from "./clone.js";
import { parseSync } from "./parse.js";
import { materializeSnapshot } from "./materialize.js";

/** @typedef {'dot'|'tail'} HistoryNodeKind */

/**
 * @typedef {{
 *   index: number,
 *   kind: HistoryNodeKind,
 *   bufferStart: number,
 *   bufferEnd: number,
 *   wire: string|null,
 *   before: unknown,
 *   after: unknown,
 *   diff: unknown,
 * }} HistoryNode
 */

export const HISTORY_NODE_KIND = Object.freeze({
  DOT: /** @type {'dot'} */ ("dot"),
  TAIL: /** @type {'tail'} */ ("tail"),
});

/**
 * @param {{
 *   snapshot?: boolean,
 *   realtime?: boolean,
 *   retainWire?: boolean,
 *   compat?: boolean|object|false,
 * }} [options]
 */
export class ParseHistory {
  constructor(options = {}) {
    /** @type {boolean} */
    this._snapshot = options.snapshot === true;
    /** @type {boolean} */
    this._realtime = options.realtime === true;
    /** @type {boolean} */
    this._retainWire = options.retainWire !== false;
    /** @type {boolean|object|false} */
    this._compat = options.compat ?? false;

    /** @type {HistoryNode[]} */
    this._nodes = [];
    /**
     * Realtime live cursor. Starts at `-1` (before first node).
     * `jumpTo(i)` requires `i > liveCursor` (forward-only along the retained sequence).
     * @type {number}
     */
    this._liveCursor = -1;
    /** @type {string|null} */
    this._sourceKey = null;
    /**
     * Maintained read-only range view for snapshot mode.
     * @type {{ from: number, to: number, nodes: HistoryNode[], json: unknown }|null}
     */
    this._rangeView = null;
  }

  /** True when either mode is on. */
  get enabled() {
    return this._snapshot || this._realtime;
  }

  get snapshotEnabled() {
    return this._snapshot;
  }

  get realtimeEnabled() {
    return this._realtime;
  }

  /** Whether per-node wire text is retained (for `jumpTo` rebuild). */
  get retainWireEnabled() {
    return this._retainWire;
  }

  get length() {
    return this._nodes.length;
  }

  /**
   * Drop all history nodes and range view (e.g. before `compactCommitted`).
   * Modes (snapshot / realtime / retainWire) stay as constructed.
   * @returns {this}
   */
  clear() {
    this._nodes.length = 0;
    this._liveCursor = -1;
    this._rangeView = null;
    return this;
  }

  /** Realtime head index (`-1` before any jump). */
  get liveCursor() {
    return this._liveCursor;
  }

  get sourceKey() {
    return this._sourceKey;
  }

  /**
   * Anytime info snapshot (safe to log / UI).
   * @returns {{
   *   snapshot: boolean,
   *   realtime: boolean,
   *   length: number,
   *   liveCursor: number,
   *   sourceKey: string|null,
   *   hasRangeView: boolean,
   *   rangeView: { from: number, to: number }|null,
   * }}
   */
  info() {
    return {
      snapshot: this._snapshot,
      realtime: this._realtime,
      length: this._nodes.length,
      liveCursor: this._liveCursor,
      sourceKey: this._sourceKey,
      hasRangeView: this._rangeView != null,
      rangeView: this._rangeView
        ? { from: this._rangeView.from, to: this._rangeView.to }
        : null,
    };
  }

  /**
   * Append a phase-boundary record. No-op when both modes are off.
   * @param {{
   *   kind?: HistoryNodeKind,
   *   bufferStart: number,
   *   bufferEnd: number,
   *   wire?: string,
   *   before: unknown,
   *   after: unknown,
   *   diff: unknown,
   * }} entry
   * @returns {HistoryNode|null}
   */
  record(entry) {
    if (!this.enabled) return null;
    const index = this._nodes.length;
    /** @type {HistoryNode} */
    const node = {
      index,
      kind: entry.kind === HISTORY_NODE_KIND.TAIL ? HISTORY_NODE_KIND.TAIL : HISTORY_NODE_KIND.DOT,
      bufferStart: entry.bufferStart | 0,
      bufferEnd: entry.bufferEnd | 0,
      wire: this._retainWire
        ? entry.wire != null
          ? String(entry.wire)
          : null
        : null,
      before: cloneJson(entry.before),
      after: cloneJson(entry.after),
      diff: cloneJson(entry.diff),
    };
    this._nodes.push(node);
    // Ingest advances the tip; realtime cursor stays until jump (forward from -1).
    this._invalidateRangeIfNeeded();
    return node;
  }

  /**
   * Snapshot: export the full node array as a **time root** (deep clone).
   * @returns {HistoryNode[]}
   */
  exportTimeRoot() {
    this._requireSnapshot("exportTimeRoot");
    return this._nodes.map((n) => cloneNode(n));
  }

  /**
   * @param {number} index
   * @returns {HistoryNode}
   */
  getNode(index) {
    return cloneNode(this._nodeAt(index));
  }

  /** @param {number} index */
  getDiff(index) {
    return cloneJson(this._nodeAt(index).diff);
  }

  /** @param {number} index */
  getBefore(index) {
    return cloneJson(this._nodeAt(index).before);
  }

  /** @param {number} index */
  getAfter(index) {
    return cloneJson(this._nodeAt(index).after);
  }

  /**
   * Snapshot: read-only compare of `after` trees at two indices.
   * @param {number} indexA
   * @param {number} indexB
   * @returns {{ a: unknown, b: unknown, indexA: number, indexB: number }}
   */
  compare(indexA, indexB) {
    this._requireSnapshot("compare");
    return {
      indexA,
      indexB,
      a: this.getAfter(indexA),
      b: this.getAfter(indexB),
    };
  }

  /**
   * Snapshot: maintain a read-only view over `[from, to]` (inclusive).
   * Re-parses concatenated retained wire when available; otherwise uses `after` of `to`.
   * @param {number} from
   * @param {number} to
   * @returns {{ from: number, to: number, nodes: HistoryNode[], json: unknown }}
   */
  viewRange(from, to) {
    this._requireSnapshot("viewRange");
    const a = this._normalizeIndex(from);
    const b = this._normalizeIndex(to);
    if (a > b) {
      throw new RangeError(`viewRange: from (${from}) > to (${to})`);
    }
    if (
      this._rangeView &&
      this._rangeView.from === a &&
      this._rangeView.to === b
    ) {
      return {
        from: a,
        to: b,
        nodes: this._rangeView.nodes.map((n) => cloneNode(n)),
        json: cloneJson(this._rangeView.json),
      };
    }

    const slice = this._nodes.slice(a, b + 1);
    const nodes = slice.map((n) => cloneNode(n));
    let json;
    const wires = slice.map((n) => n.wire).filter((w) => w != null);
    if (wires.length === slice.length) {
      const text = /** @type {string[]} */ (wires).join("");
      json = materializeSnapshot(parseSync(text, this._compat));
    } else {
      json = cloneJson(slice[slice.length - 1].after);
    }
    this._rangeView = { from: a, to: b, nodes, json };
    return {
      from: a,
      to: b,
      nodes: nodes.map((n) => cloneNode(n)),
      json: cloneJson(json),
    };
  }

  /**
   * Snapshot lifecycle: bind a source key (e.g. stream URL).
   * A **different** key releases retained nodes + range view.
   * @param {string|null|undefined} key
   * @returns {{ released: boolean, previous: string|null }}
   */
  setSource(key) {
    this._requireSnapshot("setSource");
    const next =
      key == null || key === ""
        ? null
        : typeof key === "string"
          ? key
          : String(key);
    const previous = this._sourceKey;
    if (previous != null && next != null && previous !== next) {
      this._releaseSnapshotData();
      this._sourceKey = next;
      return { released: true, previous };
    }
    if (previous != null && next == null) {
      this._releaseSnapshotData();
      this._sourceKey = null;
      return { released: true, previous };
    }
    this._sourceKey = next;
    return { released: false, previous };
  }

  /** Clear range view and all recorded nodes (snapshot release). */
  release() {
    this._requireSnapshot("release");
    this._releaseSnapshotData();
    this._sourceKey = null;
  }

  /**
   * Realtime: jump live head forward to `index`.
   * Keeps nodes `[0..index]` (positioning node retained); discards everything after.
   * Requires `index > liveCursor` (forward-only). Discarded nodes cannot be restored.
   *
   * @param {number} index
   * @returns {{
   *   index: number,
   *   kept: number,
   *   discarded: number,
   *   after: unknown,
   *   bufferEnd: number,
   *   wirePrefix: string|null,
   * }}
   */
  jumpTo(index) {
    this._requireRealtime("jumpTo");
    const i = this._normalizeIndex(index);
    if (i <= this._liveCursor) {
      throw new RangeError(
        `realtime jumpTo only moves forward (index ${i} <= liveCursor ${this._liveCursor})`,
      );
    }
    const discarded = this._nodes.length - (i + 1);
    const keptNodes = this._nodes.slice(0, i + 1);
    this._nodes = keptNodes;
    this._liveCursor = i;
    this._rangeView = null;

    const tip = keptNodes[i];
    let wirePrefix = null;
    if (this._retainWire && keptNodes.every((n) => n.wire != null)) {
      wirePrefix = keptNodes.map((n) => n.wire).join("");
    }

    return {
      index: i,
      kept: keptNodes.length,
      discarded: Math.max(0, discarded),
      after: cloneJson(tip.after),
      bufferEnd: tip.bufferEnd,
      wirePrefix,
    };
  }

  /** @param {number} index */
  canJumpTo(index) {
    if (!this._realtime) return false;
    if (!Number.isInteger(index) || index < 0 || index >= this._nodes.length) {
      return false;
    }
    return index > this._liveCursor;
  }

  // --- internals ---

  /** @param {string} api */
  _requireSnapshot(api) {
    if (!this._snapshot) {
      throw new Error(`ParseHistory.${api} requires snapshot mode`);
    }
  }

  /** @param {string} api */
  _requireRealtime(api) {
    if (!this._realtime) {
      throw new Error(`ParseHistory.${api} requires realtime mode`);
    }
  }

  /** @param {number} index */
  _normalizeIndex(index) {
    if (!Number.isInteger(index) || index < 0 || index >= this._nodes.length) {
      throw new RangeError(
        `history index out of range: ${index} (length ${this._nodes.length})`,
      );
    }
    return index;
  }

  /** @param {number} index */
  _nodeAt(index) {
    return this._nodes[this._normalizeIndex(index)];
  }

  _releaseSnapshotData() {
    this._nodes = [];
    this._liveCursor = -1;
    this._rangeView = null;
  }

  _invalidateRangeIfNeeded() {
    if (!this._rangeView) return;
    if (this._rangeView.to >= this._nodes.length) {
      this._rangeView = null;
    }
  }
}

/** @param {HistoryNode} n */
function cloneNode(n) {
  return {
    index: n.index,
    kind: n.kind,
    bufferStart: n.bufferStart,
    bufferEnd: n.bufferEnd,
    wire: n.wire,
    before: cloneJson(n.before),
    after: cloneJson(n.after),
    diff: cloneJson(n.diff),
  };
}

File: xaiop-sdk/python/src/xaiop/history.py
"""Optional parse-chain history for `.` phase boundaries."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

from .clone import clone_json
from .materialize import materialize_snapshot
from .parse import parse_sync

HistoryNodeKind = Literal["dot", "tail"]

HISTORY_NODE_KIND = {"DOT": "dot", "TAIL": "tail"}


class RangeError(ValueError):
    """Index / jump range error (Node ``RangeError`` counterpart)."""


@dataclass
class HistoryNode:
    index: int
    kind: HistoryNodeKind
    buffer_start: int
    buffer_end: int
    wire: str | None
    before: Any
    after: Any
    diff: Any


def _clone_node(n: HistoryNode) -> HistoryNode:
    return HistoryNode(
        index=n.index,
        kind=n.kind,
        buffer_start=n.buffer_start,
        buffer_end=n.buffer_end,
        wire=n.wire,
        before=clone_json(n.before),
        after=clone_json(n.after),
        diff=clone_json(n.diff),
    )


class ParseHistory:
    def __init__(
        self,
        *,
        snapshot: bool = False,
        realtime: bool = False,
        retain_wire: bool = True,
        compat: Any = False,
    ) -> None:
        self._snapshot = snapshot is True
        self._realtime = realtime is True
        self._retain_wire = retain_wire is not False
        self._compat = compat if compat is not False else False
        self._nodes: list[HistoryNode] = []
        self._live_cursor = -1
        self._source_key: str | None = None
        self._range_view: dict[str, Any] | None = None

    @property
    def enabled(self) -> bool:
        return self._snapshot or self._realtime

    @property
    def snapshot_enabled(self) -> bool:
        return self._snapshot

    @property
    def realtime_enabled(self) -> bool:
        return self._realtime

    @property
    def retain_wire_enabled(self) -> bool:
        return self._retain_wire

    @property
    def length(self) -> int:
        return len(self._nodes)

    @property
    def live_cursor(self) -> int:
        return self._live_cursor

    @property
    def source_key(self) -> str | None:
        return self._source_key

    def clear(self) -> ParseHistory:
        self._nodes.clear()
        self._live_cursor = -1
        self._range_view = None
        return self

    def info(self) -> dict[str, Any]:
        return {
            "snapshot": self._snapshot,
            "realtime": self._realtime,
            "length": len(self._nodes),
            "liveCursor": self._live_cursor,
            "sourceKey": self._source_key,
            "hasRangeView": self._range_view is not None,
            "rangeView": (
                {"from": self._range_view["from"], "to": self._range_view["to"]}
                if self._range_view
                else None
            ),
        }

    def record(self, entry: dict[str, Any]) -> HistoryNode | None:
        """Append with defensive clones (safe for external callers)."""
        return self.record_owned(
            {
                **entry,
                "before": clone_json(entry.get("before")),
                "after": clone_json(entry.get("after")),
                "diff": clone_json(entry.get("diff")),
            }
        )

    def record_owned(self, entry: dict[str, Any]) -> HistoryNode | None:
        """Append taking ownership of already-isolated trees (no extra clone).

        When ``before`` is the previous node's ``after`` (same object), adjacent
        phases share storage. Public getters still deep-clone on export.
        """
        if not self.enabled:
            return None
        index = len(self._nodes)
        before = entry.get("before")
        if self._nodes and self._nodes[-1].after is before:
            before = self._nodes[-1].after
        node = HistoryNode(
            index=index,
            kind=(
                HISTORY_NODE_KIND["TAIL"]
                if entry.get("kind") == HISTORY_NODE_KIND["TAIL"]
                else HISTORY_NODE_KIND["DOT"]
            ),
            buffer_start=int(entry["bufferStart"]),
            buffer_end=int(entry["bufferEnd"]),
            wire=(
                str(entry["wire"])
                if self._retain_wire and entry.get("wire") is not None
                else None
            ),
            before=before,
            after=entry.get("after"),
            diff=entry.get("diff"),
        )
        self._nodes.append(node)
        self._invalidate_range_if_needed()
        return node

    def export_time_root(self) -> list[HistoryNode]:
        self._require_snapshot("exportTimeRoot")
        return [_clone_node(n) for n in self._nodes]

    def get_node(self, index: int) -> HistoryNode:
        return _clone_node(self._node_at(index))

    def get_diff(self, index: int) -> Any:
        return clone_json(self._node_at(index).diff)

    def peek_diff(self, index: int) -> Any:
        """Engine emit path — no clone (do not mutate shared storage)."""
        return self._node_at(index).diff

    def get_before(self, index: int) -> Any:
        return clone_json(self._node_at(index).before)

    def get_after(self, index: int) -> Any:
        return clone_json(self._node_at(index).after)

    def peek_after(self, index: int) -> Any:
        """Adjacent-phase ``before`` sharing — no clone."""
        return self._node_at(index).after

    def compare(self, index_a: int, index_b: int) -> dict[str, Any]:
        self._require_snapshot("compare")
        return {
            "indexA": index_a,
            "indexB": index_b,
            "a": self.get_after(index_a),
            "b": self.get_after(index_b),
        }

    def view_range(self, from_index: int, to_index: int) -> dict[str, Any]:
        self._require_snapshot("viewRange")
        a = self._normalize_index(from_index)
        b = self._normalize_index(to_index)
        if a > b:
            raise RangeError(f"viewRange: from ({from_index}) > to ({to_index})")
        # Cache holds internal refs; return path always deep-clones once.
        if (
            self._range_view
            and self._range_view["from"] == a
            and self._range_view["to"] == b
        ):
            return {
                "from": a,
                "to": b,
                "nodes": [_clone_node(n) for n in self._range_view["nodes"]],
                "json": clone_json(self._range_view["json"]),
            }

        slice_nodes = self._nodes[a : b + 1]
        wires = [n.wire for n in slice_nodes]
        if all(w is not None for w in wires):
            text = "".join(str(w) for w in wires)
            json_val = materialize_snapshot(parse_sync(text, self._compat))
        else:
            json_val = slice_nodes[-1].after
        self._range_view = {
            "from": a,
            "to": b,
            "nodes": list(slice_nodes),
            "json": json_val,
        }
        return {
            "from": a,
            "to": b,
            "nodes": [_clone_node(n) for n in self._range_view["nodes"]],
            "json": clone_json(json_val),
        }

    def set_source(self, key: str | None) -> dict[str, Any]:
        self._require_snapshot("setSource")
        next_key: str | None
        if key is None or key == "":
            next_key = None
        else:
            next_key = str(key)
        previous = self._source_key
        if previous is not None and next_key is not None and previous != next_key:
            self._release_snapshot_data()
            self._source_key = next_key
            return {"released": True, "previous": previous}
        if previous is not None and next_key is None:
            self._release_snapshot_data()
            self._source_key = None
            return {"released": True, "previous": previous}
        self._source_key = next_key
        return {"released": False, "previous": previous}

    def release(self) -> None:
        self._require_snapshot("release")
        self._release_snapshot_data()
        self._source_key = None

    def jump_to(self, index: int) -> dict[str, Any]:
        self._require_realtime("jumpTo")
        i = self._normalize_index(index)
        if i <= self._live_cursor:
            raise RangeError(
                f"realtime jumpTo only moves forward (index {i} <= "
                f"liveCursor {self._live_cursor})"
            )
        discarded = len(self._nodes) - (i + 1)
        kept_nodes = self._nodes[: i + 1]
        self._nodes = kept_nodes
        self._live_cursor = i
        self._range_view = None
        tip = kept_nodes[i]
        wire_prefix = None
        if self._retain_wire and all(n.wire is not None for n in kept_nodes):
            wire_prefix = "".join(n.wire for n in kept_nodes if n.wire)
        return {
            "index": i,
            "kept": len(kept_nodes),
            "discarded": max(0, discarded),
            "after": clone_json(tip.after),
            "bufferEnd": tip.buffer_end,
            "wirePrefix": wire_prefix,
        }

    def can_jump_to(self, index: int) -> bool:
        if not self._realtime:
            return False
        if not isinstance(index, int) or index < 0 or index >= len(self._nodes):
            return False
        return index > self._live_cursor

    def _require_snapshot(self, api: str) -> None:
        if not self._snapshot:
            raise RuntimeError(f"ParseHistory.{api} requires snapshot mode")

    def _require_realtime(self, api: str) -> None:
        if not self._realtime:
            raise RuntimeError(f"ParseHistory.{api} requires realtime mode")

    def _normalize_index(self, index: int) -> int:
        if not isinstance(index, int) or index < 0 or index >= len(self._nodes):
            raise RangeError(
                f"history index out of range: {index} (length {len(self._nodes)})"
            )
        return index

    def _node_at(self, index: int) -> HistoryNode:
        return self._nodes[self._normalize_index(index)]

    def _release_snapshot_data(self) -> None:
        self._nodes = []
        self._live_cursor = -1
        self._range_view = None

    def _invalidate_range_if_needed(self) -> None:
        if self._range_view and self._range_view.get("to", 0) >= len(self._nodes):
            self._range_view = None
